csds-client/client/util.go

```diff
@@ -0,0 +1,276 @@
+type TypeResolver struct{}
+
+func (r *TypeResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// FindMessageByURL links the message type url to the specific message type
+// TODO: If there's other message type can be passed in google.protobuf.Any, the typeUrl and
+//  messageType need to be added to this method to make sure it can be parsed and output correctly
+func (r *TypeResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	switch url {
+	case "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager":
+		httpConnectionManager := envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager{}
+		return httpConnectionManager.ProtoReflect().Type(), nil
+	case "type.googleapis.com/envoy.api.v2.Cluster":
+		cluster := envoy_api_v2.Cluster{}
+		return cluster.ProtoReflect().Type(), nil
+	case "type.googleapis.com/envoy.api.v2.Listener":
+		listener := envoy_api_v2.Listener{}
+		return listener.ProtoReflect().Type(), nil
+	case "type.googleapis.com/envoy.config.filter.http.router.v2.Router":
+		router := envoy_config_filter_http_router_v2.Router{}
+		return router.ProtoReflect().Type(), nil
+	case "type.googleapis.com/envoy.api.v2.RouteConfiguration":
+		routeConfiguration := envoy_api_v2.RouteConfiguration{}
+		return routeConfiguration.ProtoReflect().Type(), nil
+	default:
+		return nil, protoregistry.NotFound
+	}
+}
+
+func (r *TypeResolver) FindExtensionByName(field protoreflect.FullName) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+func (r *TypeResolver) FindExtensionByNumber(message protoreflect.FullName, field protoreflect.FieldNumber) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// parseConfigStatus parses each xds config status to string
+func parseConfigStatus(xdsConfig []*csdspb.PerXdsConfig) []string {
+	var configStatus []string
+	for _, perXdsConfig := range xdsConfig {
+		status := perXdsConfig.GetStatus().String()
+		var xds string
+		if perXdsConfig.GetClusterConfig() != nil {
+			xds = "CDS"
+		} else if perXdsConfig.GetListenerConfig() != nil {
+			xds = "LDS"
+		} else if perXdsConfig.GetRouteConfig() != nil {
+			xds = "RDS"
+		} else if perXdsConfig.GetScopedRouteConfig() != nil {
+			xds = "SRDS"
+		}
+		if status != "" && xds != "" {
+			configStatus = append(configStatus, xds+"   "+status)
+		}
+	}
+	return configStatus
+}
+
+// printOutResponse processes response and print
+func printOutResponse(response *csdspb.ClientStatusResponse, fileName string) error {
+	if response.GetConfig() == nil || len(response.GetConfig()) == 0 {
+		fmt.Printf("No xDS clients connected.\n")
+		return nil
+	} else {
+		fmt.Printf("%-50s %-30s %-30s \n", "Client ID", "xDS stream type", "Config Status")
```

Comment:
You might want to look into using [tabwriter](https://golang.org/pkg/text/tabwriter/)

Reply:
Cool! Thanks so much! I really learn a lot of new things of Go.
However, if the flag `-output_file` is not specified, the detailed config will also be print out after these columns. Since it also use `\t` to keep the detailed config formatted and readable, perhaps using traditional formatted output will be better here to make sure it won't interfere with the format of detailed config?

Reply:
tabwriter supports stdout as well as files. You should be able to use it for the details config as well. :)

Reply:
Sorry for misunderstanding. I'll deep dive into it to use this instead in the following PRs.